File: ios/SpentyAI/SpentyAI/Features/Settings/SettingsViewModel.swift
import Foundation
import SwiftUI

@Observable
final class SettingsViewModel {

    // MARK: - State

    var settings = AppSettings()
    var currencies: [CurrencyOption] = []
    var dateFormats: [DateFormatOption] = []
    var isLoading = false
    var isSaving = false
    var errorMessage = ""
    var showError = false
    var showDeleteConfirm = false
    var showResetWarning = false
    var showResetConfirmInput = false
    var resetConfirmText = ""
    var isResetting = false
    var showResetSuccess = false
    var showSaveSuccess = false

    // Image upload state
    var isUploadingLogo = false
    var isUploadingSignature = false
    var isDeletingLogo = false
    var isDeletingSignature = false

    // MARK: - Dependencies

    private let repository: SettingsRepository
    private let authManager: AuthManager

    // MARK: - Init

    init(repository: SettingsRepository = .shared, authManager: AuthManager) {
        self.repository = repository
        self.authManager = authManager
    }

    // MARK: - Load Settings

    @MainActor
    func loadSettings() async {
        isLoading = true
        showError = false

        do {
            settings = try await repository.getSettings()
        } catch {
            handleError(error)
        }

        isLoading = false
    }

    // MARK: - Save Settings

    @MainActor
    func saveSettings() async {
        isSaving = true
        showError = false

        do {
            settings = try await repository.updateSettings(settings)
            showSaveSuccess = true
        } catch {
            handleError(error)
        }

        isSaving = false
    }

    // MARK: - Currencies

    @MainActor
    func loadCurrencies() async {
        do {
            currencies = try await repository.getCurrencies()
        } catch {
            handleError(error)
        }
    }

    // MARK: - Date Formats

    @MainActor
    func loadDateFormats() async {
        do {
            dateFormats = try await repository.getDateFormats()
        } catch {
            handleError(error)
        }
    }

    // MARK: - Logo

    @MainActor
    func uploadLogo(imageData: Data) async {
        isUploadingLogo = true
        showError = false

        do {
            let response = try await repository.uploadLogo(imageData: imageData)
            settings.logoUrl = response.resolvedUrl
        } catch {
            handleError(error)
        }

        isUploadingLogo = false
    }

    @MainActor
    func deleteLogo() async {
        isDeletingLogo = true
        showError = false

        do {
            _ = try await repository.deleteLogo()
            settings.logoUrl = nil
        } catch {
            handleError(error)
        }

        isDeletingLogo = false
    }

    // MARK: - Signature

    @MainActor
    func uploadSignature(imageData: Data) async {
        isUploadingSignature = true
        showError = false

        do {
            let response = try await repository.uploadSignature(imageData: imageData)
            settings.signatureUrl = response.resolvedUrl
        } catch {
            handleError(error)
        }

        isUploadingSignature = false
    }

    @MainActor
    func deleteSignature() async {
        isDeletingSignature = true
        showError = false

        do {
            _ = try await repository.deleteSignature()
            settings.signatureUrl = nil
        } catch {
            handleError(error)
        }

        isDeletingSignature = false
    }

    // MARK: - Reset Data

    @MainActor
    func resetData() async {
        isResetting = true
        showError = false

        do {
            _ = try await repository.resetData()
            showResetSuccess = true
            resetConfirmText = ""
        } catch {
            handleError(error)
        }

        isResetting = false
    }

    // MARK: - Delete Account

    @MainActor
    func deleteAccount() async {
        isLoading = true
        showError = false

        do {
            try await authManager.deleteAccount()
        } catch {
            handleError(error)
            isLoading = false
        }
    }

    // MARK: - Helpers

    @MainActor
    private func handleError(_ error: Error) {
        if let apiError = error as? APIError {
            errorMessage = apiError.localizedDescription
        } else {
            errorMessage = error.localizedDescription
        }
        showError = true
    }

    func dismissError() {
        showError = false
        errorMessage = ""
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/settings/SettingsViewModel.kt
package com.spentyai.app.features.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.auth.AuthManager
import com.spentyai.app.core.network.ApiResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SettingsUiState(
    val settings: AppSettings = AppSettings(),
    val currencies: List<CurrencyOption> = emptyList(),
    val dateFormats: List<DateFormatOption> = emptyList(),
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val isResetting: Boolean = false,
    val showError: Boolean = false,
    val errorMessage: String = "",
    val showSaveSuccess: Boolean = false,
    val showSignOutConfirm: Boolean = false,
    val showDeleteConfirm: Boolean = false,
    val showResetWarning: Boolean = false,
    val showResetConfirmInput: Boolean = false,
    val resetConfirmText: String = "",
    val showResetSuccess: Boolean = false,
    val isUploadingLogo: Boolean = false,
    val isUploadingSignature: Boolean = false,
    val isDeletingLogo: Boolean = false,
    val isDeletingSignature: Boolean = false
)

class SettingsViewModel(
    private val repository: SettingsRepository,
    private val authManager: AuthManager
) : ViewModel() {

    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    fun loadSettings() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, showError = false) }
            when (val result = repository.getSettings()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(settings = result.data, isLoading = false) }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            showError = true,
                            errorMessage = result.error.message ?: "Failed to load settings"
                        )
                    }
                }
            }
        }
    }

    fun saveSettings() {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, showError = false) }
            when (val result = repository.updateSettings(_uiState.value.settings)) {
                is ApiResult.Success -> {
                    _uiState.update {
                        it.copy(settings = result.data, isSaving = false, showSaveSuccess = true)
                    }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            showError = true,
                            errorMessage = result.error.message ?: "Failed to save settings"
                        )
                    }
                }
            }
        }
    }

    fun loadCurrencies() {
        viewModelScope.launch {
            when (val result = repository.getCurrencies()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(currencies = result.data) }
                }
                is ApiResult.Failure -> { /* silently fail */ }
            }
        }
    }

    fun loadDateFormats() {
        viewModelScope.launch {
            when (val result = repository.getDateFormats()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(dateFormats = result.data) }
                }
                is ApiResult.Failure -> { /* silently fail */ }
            }
        }
    }

    fun updateField(update: (AppSettings) -> AppSettings) {
        _uiState.update { it.copy(settings = update(it.settings)) }
    }

    fun showResetWarning() {
        _uiState.update { it.copy(showResetWarning = true) }
    }

    fun onResetWarningConfirm() {
        _uiState.update { it.copy(showResetWarning = false, showResetConfirmInput = true) }
    }

    fun onResetConfirmTextChanged(text: String) {
        _uiState.update { it.copy(resetConfirmText = text) }
    }

    fun resetData() {
        if (_uiState.value.resetConfirmText != "RESET") return
        viewModelScope.launch {
            _uiState.update { it.copy(isResetting = true, showResetConfirmInput = false) }
            when (repository.resetData()) {
                is ApiResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isResetting = false,
                            showResetSuccess = true,
                            resetConfirmText = ""
                        )
                    }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isResetting = false,
                            showError = true,
                            errorMessage = "Failed to reset data"
                        )
                    }
                }
            }
        }
    }

    fun showSignOutConfirm() {
        _uiState.update { it.copy(showSignOutConfirm = true) }
    }

    fun dismissSignOutConfirm() {
        _uiState.update { it.copy(showSignOutConfirm = false) }
    }

    fun showDeleteConfirm() {
        _uiState.update { it.copy(showDeleteConfirm = true) }
    }

    fun deleteAccount() {
        authManager.deleteAccount { success ->
            if (!success) {
                _uiState.update {
                    it.copy(showError = true, errorMessage = "Failed to delete account")
                }
            }
        }
    }

    fun signOut() {
        authManager.logout()
    }

    fun dismissError() {
        _uiState.update { it.copy(showError = false, errorMessage = "") }
    }

    fun dismissSaveSuccess() {
        _uiState.update { it.copy(showSaveSuccess = false) }
    }

    fun dismissResetWarning() {
        _uiState.update { it.copy(showResetWarning = false) }
    }

    fun dismissResetConfirmInput() {
        _uiState.update { it.copy(showResetConfirmInput = false, resetConfirmText = "") }
    }

    fun dismissResetSuccess() {
        _uiState.update { it.copy(showResetSuccess = false) }
        loadSettings()
    }

    fun dismissDeleteConfirm() {
        _uiState.update { it.copy(showDeleteConfirm = false) }
    }
}
